### DISCUSSION TITLE
heat_transfer_one_pipe()

### DISCUSSION THREAD

**Discussion by ladderup** [e2nIEE/pandapipes#266] [Q&A]
Hi,
I was playing around with the heat_transfer_one_pipe() pre installed network.
I wanted to compare the pandapipes headloss and flowspeed to some models we use and (sinds available in the dataframe) the openModelica.

My understanding is that the pandapipes.pipeflow(net, friction_model="colebrook", mode="all") function, first preforms a hydraulic calculation based on initial temperatures (set in junctions(tfluid_k = 293.15)). And then a temperature calculation.

Then I tried to update the junctions initial temperatures with the calculated temperatures, in a few different ways. In the uploaded script text file all different methodes can be checked.
1. original calculation (based on initial temperaturs (tfluid_k)
2. drop junctions in 'net' and create new ones with correct temperatures
3. deepcopy 'net' together with dropping and replacing junctions with updated temperatures
4. construction of a completely new network with calculated junction temperatures.

result 2 and 3 are identical to result 1.
In result 4 the headloss, flowrate (and reynolds) are updated, and match much beter with the values from openModelica and our models.

Is there a better why to recalculate the pressure drop and flowrate based on calculated temperatures? I could not with any example there the was done.

regards,
Andries Lof

PS. I'm a little bit new with these panda dataframe so i'm sorry if the code is a bit ugly
[headloss.txt](https://github.com/e2nIEE/pandapipes/files/6546569/headloss.txt)

**Comment by dcronbach** [comment]
Hallo,

You are right that the calculation is done sequentially. 

A better result can be achieved if the calculated temperatures are fed again into the hydraulic calculation, as the density and viscosity are dependent on temperature. I do not have time to look at your script, but I started to add the iterated calculation of hydraulic properties and temperature by myself. It is not reviewed by my colleagues, and some tests are missing, but I could create a separate branch for you, which contains the additional lines of code (only a few modifications are necessary). 

I think I will have time to do this on friday.

Best regards

Dennis Cronbach

**Comment by ladderup** [comment]
Dear Dennis,

That would be great.
If it's online, I will test it.
Perhaps with a small example how to initiate the iteration, I will figure
it out.

Andries

**Reply by dcronbach** [reply]
Hi,

I updated the branch "bidi_coupling" on my own pandapipes fork (dcronbach/pandapipes). You should be able to directly access this branch (let me know if this is not the case).

There are three important modifications here:

In the pipeflow module (line 103), there is a new calculation mode, called "seg". If this mode is started, the hydraulics and temperature calculation modes are started iteratively, and in each iteration, the fluid properties are updated.

The update of fluid properties is done using the function "update_properties" (line 157).

An example script can be found in the folder "non_git". You should be able to run this script without a problem.

Please note the following:

1. The code is still work in progress. I did not validate the results and there also might be logical errors I have not found so far. The function "update_properties" should also be part of the fluid module, but it was created in pipeflow.py for prototyping purposes.
2. The few tests I did so far shew that, using very small test networks, the calculation mode "seg" converges to a solution within 1 - 10 iterations. Material properties are updated.
3. That there is an update of material properties can also be seen by running the example script: If the parameter "seg_iterations" is set to anything greater than 1, properties are updated. You will notice that the flow velocity shown in the plot is not constant as it is typically for incompressible media without temperature update. 

The last point looks good for me, but, as already mentioned, I did not have time to really check the results. As far as I know, there is a new project starting in the next months which will allow us to update the heat transfer calculation modes and corresponding components.

By the way, if you decide to use the new calculation mode, I recommend copying the corresponding code to your own branch instead of replacing it with the whole branch on my fork. I am pretty sure that "bidi_coupling" is not up to date anymore...

I hope this helps

Best regards

Dennis

**Comment by ladderup** [comment]
Thank you,

I will tryout the 'seg' mode and find out how it works

Andries
